nodejs/src/stream_reader.rs

```diff
@@ -0,0 +1,208 @@
+        // this may result to a f64::MAX, but I'm not able to find a f64: From<i64> impl
+        let offset = event_data.offset_in_segment as f64;
+
+        Ok(cx.number(offset))
+    }
+
+    ///
+    /// Return the string representation.
+    ///
+    pub fn js_to_str(mut cx: FunctionContext) -> JsResult<JsString> {
+        let event_data = cx.this().downcast_or_throw::<JsBox<EventData>, _>(&mut cx)?;
+        Ok(cx.string(format!(
+            "offset {:?} data :{:?}",
+            event_data.offset_in_segment, event_data.value
+        )))
+    }
+}
+
+impl Finalize for Slice {}
+
+///
+/// This represents a segment slice which can be used to read events from a Pravega segment.
+///
+#[derive(new)]
+pub(crate) struct Slice {
+    seg_slice: Option<SegmentSlice>,
+}
+
+impl Slice {
+    fn next(&mut self) -> Option<EventData> {
+        if let Some(mut slice) = self.seg_slice.take() {
+            let next_event: Option<Event> = slice.next();
+            self.seg_slice = Some(slice);
+            next_event.map(|e| EventData {
+                offset_in_segment: e.offset_in_segment,
+                value: e.value,
+            })
+        } else {
+            info!("Empty Slice");
+            None
+        }
+    }
+
+    pub fn js_next(mut cx: FunctionContext) -> JsResult<JsBox<EventData>> {
+        // Use RefCell as self.seg_slice needs to be changed.
+        // Example at https://github.com/neon-bindings/neon/blob/0.10.0/test/napi/src/js/boxed.rs
+        let slice = cx.argument::<JsBox<RefCell<Slice>>>(0)?;
+        let event_data = slice.borrow_mut().next();
+
+        match event_data {
+            Some(event_data) => Ok(cx.boxed(event_data)),
+            None => cx.throw_error("No more data in the stream!"),
+        }
+    }
+}
+
+impl Finalize for StreamReader {}
+
+///
+/// This represents a Stream reader for a given Stream.
+/// Note: A StreamReader cannot be created directly without using the StreamReaderGroup.
+///
+#[derive(new)]
+pub(crate) struct StreamReader {
+    reader: Arc<Mutex<EventReader>>,
+    runtime_handle: Handle,
+    streams: Vec<ScopedStream>,
+}
+
+impl StreamReader {
+    // Helper method for to set the reader offline.
+    async fn reader_offline_async(&self) -> Result<(), EventReaderError> {
+        self.reader.lock().await.reader_offline().await
+    }
+
+    // Helper method for to release the segment.
+    async fn release_segment_async(&self, slice: SegmentSlice) -> Result<(), EventReaderError> {
+        self.reader.lock().await.release_segment(slice).await
+    }
+}
+
+impl StreamReader {
+    ///
+    /// Return a Slice in an asynchronous call.
+    /// The actural returned type from await will be `Promise<Slice>` aka `JsResult<JsBoxed<RefCell<Slice>>>`.
+    ///
+    /// See the tokio-fetch example for more details on how to return a Promise and await.
+    /// https://github.com/neon-bindings/examples/tree/2dbbef55f483635d0118c20c9902bf4c6faa1ecc/examples/tokio-fetch
+    ///
```

Comment:
Is this meant to be a user facing comment?

Reply:
No, all comments in rust should not be a user facing comment/doc. Users are advised to read the README.md for a tutorial and https://pravega.github.io/pravega-client-rust/nodejs/index.html for a full API reference(auto generated from ts files and contains user facing comments.) The API reference is added in https://github.com/pravega/pravega-client-rust/pull/366/files#diff-d7ecc9db1d2f1f75440ba2c3e433aec8c252e9c86a6260732400737857c71d3b and should be available in the github.io pages.